### PR TITLE
feat: Add cert validation to metrics traces

### PR DESCRIPTION
**Pull request by rakesh-garimella**
## Description

Changes proposed in this pull request (what was done and why):
- Introduce tls cert validation for traces/metrics
- Refactor tls cert validation
- The TLS Cert condition status should only be evaluated iff we have secrets with tls cert present. It should  not be evaluated in case of other secrets.

Changes refer to particular issues, PRs or documents:

- https://github.com/kyma-project/telemetry-manager/issues/930
- https://github.com/kyma-project/telemetry-manager/issues/971
## Traceability
- [x] The PR is linked to a GitHub issue.
- [ ] New features have a milestone set.
- [ ] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub issue has a respective `area` and `kind` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.
- [ ] Adjusted the documentation if the change is user-facing.
- [x] The feature is unit-tested
- [ ] The feature is e2e-tested

### CODE REVIEW

File: apis/telemetry/v1alpha1/secret_refs.go
Comment (skhalash) [inline apis/telemetry/v1alpha1/secret_refs.go:1]: Can we move this code to the reconciler packages and make the funcs unexported. No need to add it to the API pkg.

File: internal/conditions/tlsCertConditions.go
Comment (skhalash) [inline internal/conditions/tlsCertConditions.go:1]: Please fix the file name

File: internal/reconciler/logpipeline/reconciler.go [hunk 1]
```diff
@@ -350,45 +350,15 @@ func getFluentBitPorts() []int32 {
 	}
 }
 
-func getTLSCertValidationResult(ctx context.Context, pipeline *telemetryv1alpha1.LogPipeline, validator TLSCertValidator, client client.Client) cert.TLSCertValidationResult {
+func getTLSCertValidationResult(ctx context.Context, pipeline *telemetryv1alpha1.LogPipeline, validator TLSCertValidator) tlscert.TLSCertValidationResult {
 	if pipeline.Spec.Output.HTTP == nil || (pipeline.Spec.Output.HTTP.TLSConfig.Cert == nil && pipeline.Spec.Output.HTTP.TLSConfig.Key == nil) {
-		return cert.TLSCertValidationResult{
+		return tlscert.TLSCertValidationResult{
```
Comment (skhalash) [inline internal/reconciler/logpipeline/reconciler.go:355]: Can we avoid returning a dummy value here?

File: internal/reconciler/logpipeline/status.go [hunk 2]
```diff
@@ -102,44 +99,21 @@ func (r *Reconciler) setAgentHealthyCondition(ctx context.Context, pipeline *tel
...
-		message = fmt.Sprintf(conditions.MessageForLogPipeline(reason), certValidationResult.Validity.Format(time.DateOnly))
-	}
-
 	if secretref.ReferencesNonExistentSecret(ctx, r.Client, pipeline) {
 		status = metav1.ConditionFalse
 		reason = conditions.ReasonReferencedSecretMissing
 		message = conditions.MessageForLogPipeline(reason)
 	}
 
+	if !secretref.ReferencesNonExistentSecret(ctx, r.Client, pipeline) && pipeline.TLSCertAndKeyExist() {
```
Comment (skhalash) [inline internal/reconciler/logpipeline/status.go:110]: Why are we calling secretref.ReferencesNonExistentSecret twice?

File: internal/conditions/tlsCertConditions.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+func EvaluateTLSCertCondition(certValidationResult tlscert.TLSCertValidationResult) (status metav1.ConditionStatus, reason, message string) {
+	status = metav1.ConditionTrue
+	reason = ReasonConfigurationGenerated
+	message = MessageForLogPipeline(reason)
+
+	if !certValidationResult.CertValid {
+		status = metav1.ConditionFalse
+		reason = ReasonTLSCertificateInvalid
+		message = fmt.Sprintf(MessageForLogPipeline(reason), certValidationResult.CertValidationMessage)
```
Comment (skhalash) [inline internal/conditions/tlsCertConditions.go:22]: Why always use `MessageForLogPipeline`?

File: internal/tlscert/tls_cert_validator.go [hunk 4]
```diff
@@ -21,16 +28,36 @@ type TLSCertValidationResult struct {
...
 }
 
-func (tcv *TLSCertValidator) ValidateCertificate(certPEM []byte, keyPEM []byte) TLSCertValidationResult {
+func New(client client.Client) *TLSCertValidator {
+	return &TLSCertValidator{
+		client: client,
+	}
+}
+
+func (tcv *TLSCertValidator) ResolveAndValidateCertificate(ctx context.Context, certPEM *telemetryv1alpha1.ValueType, keyPEM *telemetryv1alpha1.ValueType) TLSCertValidationResult {
```
Comment (skhalash) [inline internal/tlscert/tls_cert_validator.go:37]: IMHO, ValidateCertificate would be still a good name. ResolveAndValidate seems to be too detailed.

File: docs/user/resources/04-tracepipeline.md [hunk 5]
```diff
@@ -147,13 +147,17 @@ The status of the TracePipeline is determined by the condition types `GatewayHea
...
+| ConfigurationGenerated | True             | ConfigurationGenerated  |                                                                                      |
+| ConfigurationGenerated | False            | ReferencedSecretMissing | One or more referenced Secrets are missing                                           |
+| ConfigurationGenerated | False            | MaxPipelinesExceeded    | Maximum pipeline count limit exceeded                                                |
+| ConfigurationGenerated | False            | InvalidTLSCert          | TLS certificate invalid                                                              |
+| ConfigurationGenerated | False            | InvalidTLSPrivateKey    | TLS private key invalid                                                              |
+| ConfigurationGenerated | False            | ExpiredTLSCert          | TLS certificate expired on YYYY-MM-DD                                                |
+| ConfigurationGenerated | True             | TLSCertAboutToExpire    | TLS certificate is about to expire, configured certificate is valid until YYYY-MM-DD |
```
Comment (skhalash) [inline docs/user/resources/04-tracepipeline.md:160]: Can we unify those reasons by putting TLSCert always at the beginning? @NHingerl what is your opinion?

**Review by skhalash — CHANGES_REQUESTED** [review]

File: internal/conditions/tlsCertConditions.go (hunk 3 above)
Comment (rakesh-garimella) [reply internal/conditions/tlsCertConditions.go:22]: fixed

File: internal/conditions/tlsCertConditions.go
Comment (rakesh-garimella) [reply internal/conditions/tlsCertConditions.go:1]: fixed

File: internal/tlscert/tls_cert_validator.go (hunk 4 above)
Comment (rakesh-garimella) [reply internal/tlscert/tls_cert_validator.go:37]: reverted

File: internal/reconciler/logpipeline/reconciler.go (hunk 1 above)
Comment (rakesh-garimella) [reply internal/reconciler/logpipeline/reconciler.go:355]: fixed

File: apis/telemetry/v1alpha1/secret_refs.go
Comment (rakesh-garimella) [reply apis/telemetry/v1alpha1/secret_refs.go:1]: fixed

File: internal/reconciler/logpipeline/status.go (hunk 2 above)
Comment (rakesh-garimella) [reply internal/reconciler/logpipeline/status.go:110]: fixed

File: internal/conditions/conditions.go [hunk 6]
```diff
@@ -117,6 +117,13 @@ func MessageForMetricPipeline(reason string) string {
 	return message(reason, metricPipelineMessages)
 }
 
+func CommonMessages(reason string) string {
```
Comment (skhalash) [inline internal/conditions/conditions.go:120]: Is it used anywhere except this package? If not, please make it unexported.

File: docs/user/resources/04-tracepipeline.md
Comment (skhalash) [inline docs/user/resources/04-tracepipeline.md:1]: Please, update the docs

File: internal/reconciler/telemetry/metric_components_checker.go [hunk 7]
```diff
@@ -79,13 +86,17 @@ func (m *metricComponentsChecker) firstUnhealthyPipelineReason(pipelines []telem
...
 func (m *metricComponentsChecker) determineConditionStatus(reason string) metav1.ConditionStatus {
-	if reason == conditions.ReasonNoPipelineDeployed || reason == conditions.ReasonMetricComponentsRunning {
+	if reason == conditions.ReasonNoPipelineDeployed || reason == conditions.ReasonMetricComponentsRunning || reason == conditions.ReasonTLSCertificateAboutToExpire {
 		return metav1.ConditionTrue
 	}
 	return metav1.ConditionFalse
 }
 
 func (m *metricComponentsChecker) createMessageForReason(pipelines []telemetryv1alpha1.MetricPipeline, reason string) string {
+	tlsAboutExpireMassage := m.checkTLSCertificateMessage(pipelines)
```
Comment (skhalash) [inline internal/reconciler/telemetry/metric_components_checker.go:96]:
```suggestion
	tlsAboutExpireMessage := m.checkTLSCertificateMessage(pipelines)
```

File: internal/reconciler/telemetry/metric_components_checker.go [hunk 8]
```diff
@@ -109,3 +120,13 @@ func (m *metricComponentsChecker) addReasonPrefix(reason string) string {
 	}
 	return reason
 }
+
+func (m *metricComponentsChecker) checkTLSCertificateMessage(pipelines []telemetryv1alpha1.MetricPipeline) string {
```
Comment (skhalash) [inline internal/reconciler/telemetry/metric_components_checker.go:124]: check implies that a boolean is to be returned
```suggestion
func (m *metricComponentsChecker) firstTLSCertificateMessage(pipelines []telemetryv1alpha1.MetricPipeline) string {
```

File: internal/reconciler/telemetry/trace_components_checker.go [hunk 9]
```diff
@@ -79,13 +86,18 @@ func (t *traceComponentsChecker) firstUnhealthyPipelineReason(pipelines []teleme
...
 func (t *traceComponentsChecker) determineConditionStatus(reason string) metav1.ConditionStatus {
-	if reason == conditions.ReasonNoPipelineDeployed || reason == conditions.ReasonTraceComponentsRunning {
+	if reason == conditions.ReasonNoPipelineDeployed || reason == conditions.ReasonTraceComponentsRunning || reason == conditions.ReasonTLSCertificateAboutToExpire {
 		return metav1.ConditionTrue
 	}
 	return metav1.ConditionFalse
 }
 
 func (t *traceComponentsChecker) createMessageForReason(pipelines []telemetryv1alpha1.TracePipeline, reason string) string {
+	tlsAboutExpireMassage := t.checkTLSCertificateMessage(pipelines)
```
Comment (skhalash) [inline internal/reconciler/telemetry/trace_components_checker.go:96]:
```suggestion
	tlsAboutExpireMessage := t.checkTLSCertificateMessage(pipelines)
```

File: internal/reconciler/telemetry/trace_components_checker.go [hunk 10]
```diff
@@ -107,3 +119,13 @@ func (t *traceComponentsChecker) addReasonPrefix(reason string) string {
 	}
 	return reason
 }
+
+func (t *traceComponentsChecker) checkTLSCertificateMessage(pipelines []telemetryv1alpha1.TracePipeline) string {
```
Comment (skhalash) [inline internal/reconciler/telemetry/trace_components_checker.go:123]:
```suggestion
func (t *traceComponentsChecker) findTLSCertificateMessage(pipelines []telemetryv1alpha1.TracePipeline) string {
```

File: internal/tlscert/tls_cert_validator.go [hunk 11]
```diff
@@ -21,16 +28,36 @@ type TLSCertValidationResult struct {
...
 	result := TLSCertValidationResult{
 		CertValid:       true,
 		PrivateKeyValid: true,
 		Validity:        time.Now().Add(time.Hour * 24 * 365),
 	}
 
+	certData, err := resolveValue(ctx, tcv.client, *certPEM)
+
+	if err != nil {
+		result.CertValid = false
```
Comment (skhalash) [inline internal/tlscert/tls_cert_validator.go:47]: Let's maybe set CertValidationMessage?

File: internal/tlscert/tls_cert_validator.go [hunk 12]
```diff
@@ -21,16 +28,36 @@ type TLSCertValidationResult struct {
...
+		return result
+	}
+
+	keyData, err := resolveValue(ctx, tcv.client, *keyPEM)
+
+	if err != nil {
+		result.PrivateKeyValid = false
```
Comment (skhalash) [inline internal/tlscert/tls_cert_validator.go:54]: Let's maybe set KeyValidationMessage?

**Review by skhalash — CHANGES_REQUESTED** [review]

**Pull request closed without merging**